torch/_dynamo/eval_frame.py

```diff
@@ -645,6 +648,55 @@ def __reduce__(self):
         return (self.__class__, ())
 
 
+class CompileEnabledContext:
+    def __init__(self, enabled):
+        self.enabled = enabled
+
+    def __call__(self, fn):
+        # This code is a simplified version of _TorchDynamoContext.__call__.
+        # We just want to enable/disable the eval frame callback mechanism, so we
+        # don't have to check trace_rules or create any wrapper.
+        # But we still need to return a proper callable/wrapped module/wraped class.
+        if isinstance(fn, torch.nn.Module):
+            mod = fn
+            new_mod = OptimizedModule(mod, self)
+            return new_mod
+
+        if inspect.isclass(fn):
+            # User has wrapped the class with enable/disable decorator. Apply to init/call method.
+            cls_obj = fn
+            # Disable on init is useful for reconstruction of bytecodes where we
+            # want to prevent Dynamo from tracing into the init function. Check
+            # test_reconstruction in test_model_output.py.
+            cls_obj.__init__ = self(cls_obj.__init__)
+            cls_obj.__call__ = self(cls_obj.__call__)
+            if issubclass(cls_obj, torch.nn.Module):
+                # NN module variable tracker directly inlines the _call_impl. Enable/disable it.
+                cls_obj._call_impl = self(cls_obj._call_impl)
+            return cls_obj
+
+        assert callable(fn)
+
+        @functools.wraps(fn)
+        def _fn(*args, **kwargs):
+            from torch._C._dynamo.eval_frame import set_eval_frame_callback_enabled
+
+            prior = set_eval_frame_callback_enabled(self.enabled)
+            try:
+                return fn(*args, **kwargs)
+            finally:
+                set_eval_frame_callback_enabled(prior)
+
+        if self.enabled:
+            # if dynamo is already enabled, and we are inlining, directly trace fn without going through _fn
+            _fn._torchdynamo_inline = fn  # type: ignore[attr-defined]
+        else:
+            # if dynamo is already enabled, use _torchdynamo_disable to skip going through _fn
+            _fn._torchdynamo_disable = True  # type: ignore[attr-defined]
+
+        return _fn
```

Comment:
This seems way more complicated than I expected it to be. In particular, why do you need to apply OptimizedModule?

Reply:
I get a lot of "copy paste" energy here (maybe a lot of this code was preexisting?) But then, why is it being copy pasted?

Reply:
> why do you need to apply OptimizedModule
So that the result of enable/disable can be a module if the original function is a module. e.g.
```python
mod1.layer = torch.compiler.enable(mod2)
```

> why is it being copy pasted?
Copy pasted from the original `DisableContext` so that I can preserve a copy to alleviate BC-breaking concerns. I could merge these 2 (`OldDisableContext` and `CompileEnabledContext`) together.

Reply:
I think I definitely prefer having all of the OptimizedModule wrapping logic together, although I do think there's some implementer's discretion here.

Reply:
Eventually I'd like to get rid of `OldDisableContext`, so keeping a separate copy makes it easy to remove.